### DISCUSSION TITLE
`entity.populate` doesn't function when executed on a fork

### DISCUSSION THREAD

**Discussion by SpencerKaiser** [mikro-orm/mikro-orm#3333] [Q&A] [answered]
I'm assuming this is just because I don't have a strong understanding of the Identity Map/Request Context, but can someone explain what this is all about?
``` typescript
const user = await em.findOne(User, { ... });
await em.populate(user, ['something']);
// something is populated

// If I `.fork` the em......
const user = await em.findOne(User, { ... });
await em.fork().populate(user, ['something']);
// something is just the id, no request is even made to the DB
```

I ended up using this to make it work (I don't have access to the original `em` at the point where I need it):
```typescript
await user.something.unwrap().init();
user.something.populated(true);
```

That feels pretty gross... is there a better way to do this? I couldn't seem to find any docs on what the `init` method even does, but the sql it executes is correct 😬

**Comment by B4nan** [comment, marked as answer]
`await em.fork().populate(user, ['something']);` operates in a different context - a new one you just created by doing `em.fork()`. That context is clean, no entities inside, so populating to-one relations can't work - as that works by initializing existing entity references in the identity map. There are none, so no queries are made.

So the question is, why do you fork there?

edit: you could try `em.fork({ clear: false }).populate(...)`, that will copy the contents of identity map to the fork

**Reply by SpencerKaiser** [reply]
Gotcha that makes sense

> So the question is, why do you fork there?
We have middleware in the app that checks to make sure the user exists and we do so by querying. We don't want to inadvertently populate relations within handlers, so we do that query on a fork and then pass a clean entity manager to the handler. I guess we could pass the `originalEntityManager` along with the request so we can use it for situations like this if needed...

> you could try em.fork({ clear: false }).populate(...), that will copy the contents of identity map to the fork
It's helpful to know this exists, but yea, we want to make sure the child gets a clean context/identify map so we don't accidentally populate stuff. Make sense?

Lastly, any idea why `init` + `populated(true)` works but `load` doesn't? Is it also because it's dependent on the context?

I think my best approach is to pass that original entity manager along with the request for use-cases like this (although this is the _**first**_ time it's come up...) unless you have another recommendation 😄

**Reply by B4nan** [reply]
You can always get the EM from entity instance (the one that is managing it, the one used to load it). Although its "private api", but I dont think it will ever change:

```ts
const em = wrap(entity, true).__em;
```

This is the same EM used for the `init` method. As opposed to the clean fork if you do `em.fork().populate()`.

**Reply by SpencerKaiser** [reply]
Ahhhhh so `init` uses the original em, that makes much more sense... but why is the `populated(true)` necessary? And any docs on `init`? I couldn't find any...

Also thank you (as always), this has been tremendously helpful

**Reply by B4nan** [reply]
Docs are a bit sparse, but [here](https://mikro-orm.io/docs/entity-helper). Originally the `EntityHelper` class was responsible for many things, now its split into several classes, so thats why the weird URL.

> but why is the populated(true) necessary? 

Because `init` marks the entity as `__lazyInitialized`. I believe it was to ensure the serialized response is stable. But seeing we have the `populated` parameter in `init`, I am not 100% sure whats the full story behind it. Serialization is quite a complex task here, given we have one entity instance that can be on multiple places in the graph.

**Reply by B4nan** [reply]
Like the goal was definitely to not blindly populate things when you use `init`, I know that for sure. Just not sure about the implementation. 

E.g. if you return an author with a list of books, and you would call `init` on some relation of a single book - you most probably just wanted to have that relation loaded on runtime, and not in the API response. As opposed to using populate hints or `em.populate()`.

**Reply by SpencerKaiser** [reply]
Awesome, thanks for the added context!!! I appreciate you and the project!
